### PR TITLE
`myposts.vue` - rewrite to use the composition API

### LINKED ISSUE #1: Convert myposts.vue to script setup
`myposts.vue` uses the Vue3 options API, with a `setup()` method.  We should convert it to the `script setup` model, including watches and computed properties.

### PR DESCRIPTION
**Pull request by smellyshovel**
Close #1

### CODE REVIEW

File: pages/myposts.vue [hunk 1]
```diff
@@ -15,7 +15,11 @@
             :chats="me.expectedchats"
           />
           <div v-if="!me" class="d-flex justify-content-center mt-4 flex-wrap">
-            <b-button variant="primary" size="lg" @click="forceLogin">
+            <b-button
+              variant="primary"
+              size="lg"
+              @click="authStore.forceLogin = true"
```
Comment (smellyshovel) [inline pages/myposts.vue:21]: The method was only used once - **inlined**

File: pages/myposts.vue [hunk 2]
```diff
@@ -130,13 +138,16 @@
                       <b-button
                         variant="secondary"
                         title="Show old WANTEDs"
-                        @click="toggleOldWanted"
+                        @click="showOldWanteds = !showOldWanteds"
```
Comment (smellyshovel) [inline pages/myposts.vue:141]: Same story as with the offers (see above).

File: pages/myposts.vue [hunk 3]
```diff
@@ -244,314 +255,250 @@
...
-import { useGroupStore } from '../stores/group'
 import { useAuthStore } from '../stores/auth'
 import { useMessageStore } from '../stores/message'
 import { useMiscStore } from '../stores/misc'
-import { useComposeStore } from '../stores/compose'
 import { useSearchStore } from '../stores/search'
-import VisibleWhen from '~/components/VisibleWhen'
 import { buildHead } from '~/composables/useBuildHead'
+
+import VisibleWhen from '~/components/VisibleWhen'
```
Comment (smellyshovel) [inline pages/myposts.vue:269]: Had to get rid of all the dynamically-imported components: Nuxt seems to break when used in the script setup. Can research more on the topic, or we can leave it as this.

FYI, in other components which have already been rewritten to use script setup, I don't see any dynamically-imported components.

File: pages/myposts.vue [hunk 4]
```diff
@@ -244,314 +255,250 @@
     </b-container>
   </client-only>
 </template>
-<script>
+
+<script setup>
 import { useRoute } from 'vue-router'
 import pluralize from 'pluralize'
 import dayjs from 'dayjs'
-import { useGroupStore } from '../stores/group'
```
Comment (smellyshovel) [inline pages/myposts.vue:251]: Group Store, Compose Store and some others seem not to be used really - **removed**.

File: pages/myposts.vue [hunk 5]
```diff
@@ -244,314 +255,250 @@
...
-    const messageStore = useMessageStore()
-    const groupStore = useGroupStore()
-    const miscStore = useMiscStore()
-    const composeStore = useComposeStore()
-    const searchStore = useSearchStore()
-
-    // We want this to be our next home page.
-    const existingHomepage = miscStore.get('lasthomepage')
-
-    if (existingHomepage !== 'myposts') {
```
Comment (smellyshovel) [inline pages/myposts.vue:311]: This block is removed:

1. I don't understand what it's for
2. If it's needed for some later time, let's re-add it at that later time. No need to pullute the existing code base.

File: pages/myposts.vue [hunk 6]
```diff
@@ -244,314 +255,250 @@
...
+
+        miscStore.set({
+          key: 'lastdonationask',
+          value: new Date().getTime(),
+        })
+      }
     }
-  },
-  computed: {
-    postcode() {
```
Comment (smellyshovel) [inline pages/myposts.vue:365]: Hasn't been used anywhere - **removed**

File: pages/myposts.vue [hunk 7]
```diff
@@ -244,314 +255,250 @@
...
+    null,
+    {
+      class: 'overflow-y-scroll',
     }
+  )
+)
 
-    // We might have parameters from just having posted.
-    const newuser = route.params.newuser
-    const newpassword = route.params.newpassword
```
Comment (smellyshovel) [inline pages/myposts.vue:320]: Don't see any usage of these two - **removed**.

**Comment by smellyshovel** [comment]
@edwh ready for review.

File: pages/myposts.vue (hunk 1 above)
Comment (edwh) [reply pages/myposts.vue:21]: I understand why you would do this, but I prefer separate methods please.

File: pages/myposts.vue (hunk 3 above)
Comment (edwh) [reply pages/myposts.vue:269]: I think the components importing works better in this model, so the dynamic import is not needed.

File: pages/myposts.vue (hunk 5 above)
Comment (edwh) [reply pages/myposts.vue:311]: Some people like the home page (the page we load when they visit the site) to be My Posts, some like it to be Browse, some people like it to be ChitChat.

So we store this persistently on the client - see `misc.js` for the persisting strategy.  Each of these potential home pages updates that value.  Then in `pages/index.vue` we read it and then route / to the appropriate page, which means you are taken to the last of these pages which you visited.

So please add this back in.

File: pages/myposts.vue (hunk 6 above)
Comment (edwh) [reply pages/myposts.vue:365]: Agree.

File: pages/myposts.vue (hunk 4 above)
Comment (edwh) [reply pages/myposts.vue:251]: Agree.

File: pages/myposts.vue [hunk 8]
```diff
@@ -78,13 +85,14 @@
...
                     >
+                      HERE
                       <MyMessage
                         :id="message.id"
                         :show-old="showOldOffers"
                         :expand="expand"
                       />
                     </div>
-                    <infinite-loading
+                    <InfiniteLoading
```
[the remark this answers is not in the document]
Comment (edwh) [reply pages/myposts.vue:95]: Yes, it is please.

**Comment by edwh** [comment]
After trying this, it's going to be easier for me to review if you push a branch to the main repo to create your PR.  So I've given you write permissions.  But for this PR we'll use your fork.

There's some `HERE` debug in the template - please remove.

For `askmodal` you have `ref()` - I'd prefer `ref(false)` to be explicit.

For me the button to show old WANTEDs isn't showing anything, but it does on the live site.  So that might be a bug - can you investigate?

This gives us a `setup` script which is quite long - so for example the donation ask logic is partly in `onMounted` and partly at the bottom of the script.  I find this a problem with the composition approach - it's hard to find related bits of logic.  Do you think we could/should split this page to make it simpler?

**Comment by smellyshovel** [comment]
> After trying this, it's going to be easier for me to review if you push a branch to the main repo to create your PR. So I've given you write permissions. But for this PR we'll use your fork.

Got you.

> There's some HERE debug in the template - please remove.

Thanks for noticing. Done.

> For askmodal you have ref() - I'd prefer ref(false) to be explicit.

Strongly disagree. If we were using typescript, such an approach wouldn't even compile at all. All the examples in Vue docs (https://vuejs.org/guide/essentials/template-refs.html) show the use of explicit `null`, but not `false`. In my case it's undefined, which doesn't make much difference with `null`, but I believe that `false` would be too counterintuitive. It would look like it's some kind of a flag variable, which is later toggled to `true` (instead of being resolved to an element/component). If you're 100% sure you want to do that, please, let me know. But I think we should better leave it as it is.

> This gives us a setup script which is quite long - so for example the donation ask logic is partly in onMounted and partly at the bottom of the script. I find this a problem with the composition approach - it's hard to find related bits of logic. Do you think we could/should split this page to make it simpler?

Your point is clear to me, it's just that I didn't understand that `onMounted` hold some logic related to donation ask. I'll try to figure out what the code's doing and make appropriate changes to it.

> For me the button to show old WANTEDs isn't showing anything, but it does on the live site. So that might be a bug - can you investigate?

Not too surprised since I didn't test it. Old WANTEDs - how do I get them for my account? Do I need someone to give me an item a wanted first?

File: pages/myposts.vue (hunk 1 above)
Comment (smellyshovel) [reply pages/myposts.vue:21]: Reverted.

File: pages/myposts.vue [hunk 9]
```diff
@@ -41,13 +45,16 @@
                       <b-button
                         variant="secondary"
                         title="Show old OFFERs"
-                        @click="toggleOldOffer"
+                        @click="showOldOffers = !showOldOffers"
```
[the remark this answers is not in the document]
Comment (smellyshovel) [reply pages/myposts.vue:48]: Reverted.

File: pages/myposts.vue (hunk 2 above)
Comment (smellyshovel) [reply pages/myposts.vue:141]: Reverted.

File: pages/myposts.vue (hunk 5 above)
Comment (smellyshovel) [reply pages/myposts.vue:311]: Thanks for the explanations! Reverted.

**Comment by smellyshovel** [comment]
> For me the button to show old WANTEDs isn't showing anything, but it does on the live site. So that might be a bug - can you investigate?

> Not too surprised since I didn't test it. Old WANTEDs - how do I get them for my account? Do I need someone to give me an item a wanted first?

Are you sure it doesn't work? For me it does:

https://github.com/Freegle/iznik-nuxt3/assets/9782236/0d1a7a57-b173-4cf6-b8f2-0ae5b852909d

**Comment by smellyshovel** [comment]
> This gives us a setup script which is quite long - so for example the donation ask logic is partly in onMounted and partly at the bottom of the script. I find this a problem with the composition approach - it's hard to find related bits of logic. Do you think we could/should split this page to make it simpler?

Regarding this one. I don't see that the `onMounted` hold any logic other than the one responsible for donations asks. So as for me, there's no point in splitting this `onMounted` into multiple ones. However, what I did instead, is that I moved this `onMounted` closer to the code related to WANTEDs.

Could you please elaborate more on the issue? Where is that line in the callback which separates WANTEDs-related logic from some other logic?

**Comment by edwh** [comment]
> > For askmodal you have ref() - I'd prefer ref(false) to be explicit.
> 
> Strongly disagree. If we were using typescript, such an approach wouldn't even compile at all. All the examples in Vue docs (https://vuejs.org/guide/essentials/template-refs.html) show the use of explicit `null`, but not `false`. In my case it's undefined, which doesn't make much difference with `null`, but I believe that `false` would be too counterintuitive. It would look like it's some kind of a flag variable, which is later toggled to `true` (instead of being resolved to an element/component). If you're 100% sure you want to do that, please, let me know. But I think we should better leave it as it is.

You're completely right.  I read this as being a flag, but it isn't.

**Comment by edwh** [comment]
> Could you please elaborate more on the issue? Where is that line in the callback which separates WANTEDs-related logic from some other logic?

Sorry, I wasn't clear.  It's not about WANTEDs specifically, it's about the general length of the setup script and whether we could reduce and simplify it.  Some ways we could shorten it:
- I think the `lastdonationask` logic could be moved into `DonationAskModal`.  That also affects `[[action]].vue` which isn't currently using the same logic, but should be.  That would shorten `setup` in `myposts.vue`.
- We could have a separate component for the list, which we then reuse for both OFFERs/WANTEDs.  That would reduce some duplicate code.
- The Searches list could be a separate component.
- The `lasthomepage` logic about saving the home page could be in a composable so that it was shared between the pages that use it rather than duplicated.  Maybe this could even be in the layout.

I'm not sure whether to put effort into this or not.  What do you think, since this is the first time you've seen the code?  JetBrains has some function to extract components, though it didn't work that well last time I tried it.

> > For me the button to show old WANTEDs isn't showing anything, but it does on the live site. So that might be a bug - can you investigate?
> 
> > Not too surprised since I didn't test it. Old WANTEDs - how do I get them for my account? Do I need someone to give me an item a wanted first?
> 
> Are you sure it doesn't work? For me it does:

I think this is because inside the `wanteds` computed you're using `showOldOffers` - should be `showOldWanted`.

Also, I think the `forceLogin` logic, the top-level `v-if="me"` and the corresponding part of the template is obsolete now.  This is old code which predates the `login` layout now works.  So this can be removed, but please test the logged out case.

**Comment by smellyshovel** [comment]
@edwh I agree with all the points, but I think it'd be better to make this an effort of another issue/PR. This one is about converting from options to composition (which is done), in another one we could continue with refactoring the component. What do you think?

**Comment by smellyshovel** [comment]
@edwh sure, will do. What about this one though? Ragarding the initial task (Options API -> Composition API convertion), is everything addressed?

**Comment by edwh** [comment]
@smellyshovel Yes - I'll merge.  Nice to get the first one in!

**Pull request merged**